gbif-occurrence-linkage/main.py

```diff
@@ -2,49 +2,60 @@
+    When done, it will publish an annotation to annotation processing service
     """
-    consumer = KafkaConsumer(
-        os.environ.get("KAFKA_CONSUMER_TOPIC"),
-        group_id=os.environ.get("KAFKA_CONSUMER_GROUP"),
-        bootstrap_servers=[os.environ.get("KAFKA_CONSUMER_HOST")],
-        value_deserializer=lambda m: json.loads(m.decode("utf-8")),
-        enable_auto_commit=True,
-    )
-    producer = KafkaProducer(
-        bootstrap_servers=[os.environ.get("KAFKA_PRODUCER_HOST")],
-        value_serializer=lambda m: json.dumps(m).encode("utf-8"),
+    connection = pika.BlockingConnection(
+        pika.ConnectionParameters(
+            os.environ.get("RABBITMQ_HOST"),
+            credentials=pika.PlainCredentials(
+                os.environ.get("RABBITMQ_USER"),
+                os.environ.get("RABBITMQ_PASSWORD")),
+        )
     )
-    for msg in consumer:
-        try:
-            logging.info("Received message: " + str(msg.value))
-            json_value = msg.value
-            shared.mark_job_as_running(json_value.get("jobId"))
-            specimen_data = json_value.get("object")
-            result = run_api_call(specimen_data)
-            annotation_event = map_to_annotation_event(
-                specimen_data, result, json_value.get("jobId")
-            )
-            publish_annotation_event(annotation_event, producer)
-        except Exception as e:
-            logging.exception(e)
-
-
-def map_to_annotation_event(
-    specimen_data: Dict, result: Dict[str, str], job_id: str
-) -> dict:
+    channel = connection.channel()
+    channel.basic_consume(queue=os.environ.get("RABBITMQ_QUEUE"), on_message_callback=process_message, auto_ack=True)
+    channel.start_consuming()
+
+
+def process_message(channel: BlockingChannel, method: Method, properties: Properties, body: bytes) -> None:
+    """
+    Callback function to process the message from RabbitMQ. This method will be called for each message received.
+    We will first convert it to JSON and extract the accessURI.
+    Then we run Pillow to extract the image metadata and create an annotation.
+    We publish this annotation through the channel on a RabbitMQ exchange.
+    :param channel: The RabbitMQ channel, which we will use to publish the resulting annotation
+    :param method: The method used to send the message, not currently used
+    :param properties: Properties of the message, not currently used
+    :param body: The message body in bytes
+    :return:
+    """
+    try:
+        json_value = json.loads(body.decode("utf-8"))
+        shared.mark_job_as_running(json_value.get("jobId"))
+        specimen_data = json_value.get("object")
+        result = run_api_call(specimen_data)
+        annotation_event = map_to_annotation_event(specimen_data, result, json_value.get("jobId"))
+        publish_annotation_event(annotation_event, channel)
+    except Exception as e:
+        logging.exception(e)
```

Comment:
can you also send the failed message back? (See plant organ segmentation)

`send_failed_message(json_value["jobId"], str(e), channel)`

Reply:
Added